linkerd/linkerd2
# Adding edges endpoint to public API

This PR is the API portion of #2708. 

It adds an endpoint to the public API to allow us to query Prometheus for edge data, in order to display identity information for connections between Linkerd proxies.

This PR ONLY includes changes to the controller and protobuf.

The corresponding CLI code for a `linkerd edges` query is written, but I want feedback on this portion before making the second PR.

### Notes
* Currently a request only returns data for *two-sided edges* for `pods` and `deployments`. 
* A request does not return data for one-sided edges, i.e. health checks or Prometheus/tap queries. This will come after two-sided edges are correctly displayed.
* While currently, a request returns some data for other resource types supported by `linkerd stat` (for example, `authority`) it's unclear if the current results are complete/useful, I need feedback on this.
* This PR does not include a test, when I get feedback/agreement on what's returned, I'll add that.

### Prometheus querying

For each edge, we are returning:
`src name` - resource name
`dst name` - resource name
`client_id` - service account name for source
`server_id` - service account name for destination
`no_tls_reason` - reason for no identity, if none known

To get this info from Prometheus we are combining information from two different queries, an inbound and outbound query. 
We use `label_replace` in Prometheus to create a key that we can match `inbound` to `outbound` queries with, so we can populate the whole `edge` result.

Remember that we are currently filtering out edges that are one-sided (i.e. health probes, Prometheus, tap) and also edges where we cannot definitively match the inbound/outbound queries together. That will be the next step.

### To Test
1. Update protobuf: 
 If you make Protobuf changes, run:
 ```bash
bin/dep ensure
bin/protoc-go.sh
```

2. Run the controller locally:
 ```bash
bin/go-run controller/cmd/public-api -log-level debug
```
You should see `starting HTTP server on :8085`

3. Make a temporary `main.go` file that you can test API calls with, in the `/linkerd2` directory (suggested code at bottom of this PR)

4. In a separate terminal window, run 
 ```bash
bin/go-run .
```
See what data is returned.

5. If you are requesting edges between `pods` in `booksapp`, you can compare to the result from:
 ```bash
linkerd stat po --from deploy -n booksapp
linkerd stat po --to deploy -n booksapp
```
(Note, some of those results may be one-sided (i.e. a probe) so not all will appear)

=============

Suggested `main.go` file contents, you can switch out different `Namespace` and `Type` values:
```
package main

import (
	"context"
	"fmt"

	"github.com/linkerd/linkerd2/controller/api/public"
	pb "github.com/linkerd/linkerd2/controller/gen/public"
)

func main() {
	client, err := public.NewInternalClient("linkerd", "localhost:8085")
	if err != nil {
		panic(err)
	}

	rsp, err := client.Edges(context.Background(), &pb.StatSummaryRequest{
		Selector: &pb.ResourceSelection{
			Resource: &pb.Resource{
				Namespace: "booksapp",
				Type:      "pod",
			},
		},
		TimeWindow: "1m",
	})
	if err != nil {
		panic(err)
	}

	fmt.Printf("%+v\n", rsp)
}
```
